### PR TITLE
feat: extend the travel mode filter selection to set filter using the selected_as_default property 

### PR DESCRIPTION
**Pull request by jonasbrunvoll**
Closes https://github.com/AtB-AS/kundevendt/issues/17913

 ### Background

FRAM and NFK have requested that some transport mode filters should be unchecked by default in both app and travel planner web. This request highlights a need that we need to address. As the number of organizations in the OMS grows, it becomes increasingly more difficult to find on singel transport mode configuration that will work well for all orgnaizations. Therefore, each organization should be able to configure their own desired default transportModes. 

The requests from AtB, FRAM and NFK are as following:
-  **AtB:** All transport mode filters should be checked by default (7/7).
-  **NFK:** All transport mode filters except flight should be check by default (6/7).
-  **FRAM:** All transport mode filters except flight and train should be checked by default (5/7).

#### Illustrations
<details>
<summary>screenshots/video/figma</summary>

https://github.com/AtB-AS/mittatb-app/assets/59939294/afdcef3c-2d33-4a4e-b24a-94f394116a5e

https://github.com/AtB-AS/mittatb-app/assets/59939294/673697f1-cf5c-4747-b468-0601deaa6e8d

</details>

### Proposed solution
- [x] Use the property `selectedAsDefault` to determine whether a transport mode filters should be checked or unchecked by default. 
- [x] Display a message box informing the users about how many of the filters are checked, if any other combination but the default filters are checked. 
- [x] Reset to default filters when the box with filter information is crossed away. 

### Acceptence criteria 
-  The correct transport mode filters are checked by default.
-  The user is still able to specify which transport mode filters to use during the search.   

**User Story 1:**  
As a user, I want to perform a travel search with default travel mode filters.

**Checklist with 7/7 default filters:**
- [ ]  The user specifies the desired start and end destination in the search fields.
- [ ]  The green dot is not visible on the filter button.
- [ ]  The message "7 out of 7 filters selected" is not displayed.
- [ ]  The user clicks the search button.
- [ ]  The search results appear on the screen.

**Checklist with 6/7 default filters:**
- [ ]  The user specifies the desired start and end destination in the search fields.
- [ ]  The green dot is not visible on the filter button.
- [ ]  The message "6 out of 7 filters selected" is not displayed.
- [ ]  The user clicks the search button.
- [ ]  The search results appear on the screen.

**User Story 2.1:**  
As a user, I want to perform a travel search with custom travel mode filters.

**Checklist with 7/7 default filters:**

- [ ] The user specifies the desired start and end destination in the search fields.
- [ ] The user specifies the custom travel mode filter. This user wants to only travel by bus.
- [ ] The green dot is visible on the filter button.
- [ ] The message "1 out of 7 filters selected" is displayed.
- [ ] The user clicks the search button.
- [ ] The search results appear on the screen.

**Checklist with 6/7 default filters:**

- [ ] The user specifies the desired start and end destination in the search fields.
- [ ] The user specifies the custom travel mode filter. This user wants to only travel by bus.
- [ ] The green dot is visible on the filter button.
- [ ] The message "1 out of 7 filters selected" is displayed.
- [ ] The user clicks the search button.
- [ ] The search results appear on the screen.

**User Story 2.2:**  
As a user, I want to perform a travel search with custom travel mode filters.

**Checklist with 6/7 default filters:**
- [ ]  Default filters are: bus, train, ferry, car ferry, airport bus, and **other**.
- [ ]  The user specifies the desired start and end destination in the search fields.
- [ ]  The user specifies the custom travel mode filters. User wants to travel by: bus, train, ferry, car ferry, airport bus, and **flight**.
- [ ]  The green dot is visible on the filter button.
- [ ]  The message "6 out of 7 filters selected" is displayed.
- [ ]  The user clicks the search button.
- [ ]  The search results appear on the screen.

**User Story 3:**  
As a user, I want to reset the travel search from using custom filter options to default filter options.

**Checklist with 7/7 default filters:**

- [ ]  The user has previously performed a travel search with only bus as the selected travel mode filter.
- [ ]  The green dot is visible on the filter button.
- [ ]  The message "1 out of 7 filters selected" is displayed.
- [ ]  The user clears the message "1 out of 7 filters selected" --> Resets the filter to the default 8/8 selection.
- [ ]  The green dot is not visible on the filter button.
- [ ]  The message "1 out of 7 filters selected" is not displayed.
- [ ]  The user clicks the search button.
- [ ]  The search results appear on the screen.

**Checklist with 6/7 default filters:**
- [ ]  The user has previously performed a travel search with only bus as the selected travel mode filter.
- [ ]  The green dot is visible on the filter button.
- [ ]  The message "1 out of 7 filters selected" is displayed.
- [ ]  The user clears the message "1 out of 7 filters selected" --> Resets the filter to the default 6/7 selection.
- [ ]  The green dot is not visible on the filter button.
- [ ]  The message "1 out of 7 filters selected" is not displayed.
- [ ]  The user clicks the search button.
- [ ]  The search results appear on the screen.

**User Story 4:**  
As a user, I want to save use of custom travel mode fileters for future travel searches.

**Checklist with 7/7 default filters:**
- [ ] The user specifies the desired start and end destination in the search fields.
- [ ] The user specifies the custom travel mode filters. This time the user specifies only bus. 
- [ ] The user checks the checkbox "Save selections for future searches".
- [ ]  The green dot is visible on the filter button.
- [ ]  The message "1 out of 7 filters selected" is displayed.
- [ ]  The user clicks the search button.
- [ ]  The search results appear on the screen

**Checklist with 6/7 default filters:**
- [ ] The user specifies the desired start and end destination in the search fields.
- [ ] The user specifies the custom travel mode filters. This time the user specifies only bus. 
- [ ] The user checks the checkbox "Save selections for future searches".
- [ ]  The green dot is visible on the filter button.
- [ ]  The message "1 out of 7 filters selected" is displayed.
- [ ]  The user clicks the search button.
- [ ]  The search results appear on the screen

**User Story 5:**  
As a user, I want to remove saved custom travel mode fileters.

**Checklist with 7/7 default filters:**
- [ ] The user removes the specified custom travel mode filters. --> Resets the filter to the default 7/7 selection.
- [ ]  The green dot is not visible on the filter button.
- [ ]  The message "X out of 7 filters selected" is not displayed.
- [ ]  The user clicks the search button.
- [ ]  The search results appear on the screen

**Checklist with 6/7 default filters:**
- [ ] The user removes the specified custom travel mode filters. --> Resets the filter to the default 6/7 selection.
- [ ]  The green dot is not visible on the filter button.
- [ ]  The message "X out of 7 filters selected" is not displayed.
- [ ]  The user clicks the search button.
- [ ]  The search results appear on the screen

### CODE REVIEW

**Review by gorandalum — CHANGES_REQUESTED** [review]
In my testing the green dot logic didn't work as described in the user-stories. On a fresh install there was a green dot on the filters button if not every transport mode was selected as default.

Also there needs to be some description of backwards compatibility things to test, to be sure it works well with filters the user has previously stored.

Note that acceptance criteria and test input should be described on the issue/PR that is in the board. Since this PR is not in the board, it should be described on the connected issue.

**Comment by jonasbrunvoll** [comment]
> In my testing the green dot logic didn't work as described in the user-stories. On a fresh install there was a green dot on the filters button if not every transport mode was selected as default.
> 
> Also there needs to be some description of backwards compatibility things to test, to be sure it works well with filters the user has previously stored.
> 
> Note that acceptance criteria and test input should be described on the issue/PR that is in the board. Since this PR is not in the board, it should be described on the connected issue.

A blunder on my part to forget the logic of the green dot. Will fix it now

File: src/stacks-hierarchy/Root_TabNavigatorStack/TabNav_DashboardStack/Dashboard_TripSearchScreen/__tests__/are_non_default_filters_selected.test.ts
Comment (gorandalum) [inline src/stacks-hierarchy/Root_TabNavigatorStack/TabNav_DashboardStack/Dashboard_TripSearchScreen/__tests__/are_non_default_filters_selected.test.ts:1]: Remove 'non' from the file name? As the function is named `areDefaultFiltersSelected`.

File: src/stacks-hierarchy/Root_TabNavigatorStack/TabNav_DashboardStack/Dashboard_TripSearchScreen/__tests__/are_non_default_filters_selected.test.ts [hunk 1]
```diff
@@ -0,0 +1,25 @@
+import {TransportModeFilterOptionWithSelectionType} from '@atb/travel-search-filters';
+import {areDefaultFiltersSelected} from '../utils';
+
+describe('areDefaultFiltersSelected', () => {
+  it('should return false if input `transportModes` is undefined.', () => {
+    expect(areDefaultFiltersSelected()).toBeFalsy();
+  });
```
Comment (gorandalum) [inline src/stacks-hierarchy/Root_TabNavigatorStack/TabNav_DashboardStack/Dashboard_TripSearchScreen/__tests__/are_non_default_filters_selected.test.ts:7]: Could also add a test for empty array.

File: src/stacks-hierarchy/Root_TabNavigatorStack/TabNav_DashboardStack/Dashboard_TripSearchScreen/utils.ts [hunk 2]
```diff
@@ -107,3 +116,10 @@ export const sanitizeSearchTime = (searchTime: SearchTime) =>
   searchTime.option === 'now'
     ? {...searchTime, date: new Date().toISOString()}
     : searchTime;
+
+export const areDefaultFiltersSelected = (
+  transportModes?: TransportModeFilterOptionWithSelectionType[],
+): boolean => {
+  if (!transportModes) return false;
+  return !transportModes.some((tm) => tm.selectedAsDefault !== tm.selected);
```
Comment (gorandalum) [inline src/stacks-hierarchy/Root_TabNavigatorStack/TabNav_DashboardStack/Dashboard_TripSearchScreen/utils.ts:124]: I think `transportModes.every((tm) => tm.selectedAsDefault === tm.selected)` is quite a bit easier to read than using `some` which needs two negations in the statement. And I think they should end up the same 😅 At least we have some tests to check it.

**Review by gorandalum — APPROVED** [review]
Good work! Some minor comments that can be taken a look at before merging :)
